services/lambda-pr-status-labeler/PRStatusBot.py

```diff
@@ -0,0 +1,298 @@
+    def _parse_reviews(self, pr):
+        """
+        This method parses through the reviews of the PR and returns count of
+        4 states: Approved reviews, Comment reviews, Requested Changes reviews
+        and Dismissed reviews
+        :param pr
+        """
+        approved_count, requested_changes_count, comment_count = 0, 0, 0
+        for review in pr.get_reviews():
+            if review.state == 'APPROVED':
+                approved_count += 1
+            elif review.state == 'CHANGES_REQUESTED':
+                requested_changes_count += 1
+            elif review.state == 'COMMENTED':
+                comment_count += 1
+            elif review.state == 'DISMISSED':
+                dismissed_count += 1
+            else:
+                logging.error(f'Unknown review state {review.state}')
+        return approved_count, requested_changes_count, comment_count, dismissed_count
+
+    def _label_pr_based_on_status(self, combined_status_state, pull_request_obj):
+        """
+        This method checks the CI status of the specific commit of the PR
+        and it labels the PR accordingly
+        :param combined_status_state
+        :param pull_request_obj
+        """
+        # pseudo-code
+        # if WIP in title or PR is draft or CI failed:
+        #   pr-work-in-progress
+        # elif CI has not started yet or CI is in progress:
+        #   pr-awaiting-testing
+        # else: # CI passed checks
+        #   if pr has at least one approval and no request changes:
+        #       pr-awaiting-merge
+        #   elif pr has no review or all reviews have been dismissed/re-requested:
+        #       pr-awaiting-review
+        #   else: # pr has a review that hasn't been dismissed yet no approval
+        #       pr-awaiting-response
+
+        # combined status of PR can be 1 of the 3 potential states
+        # https://developer.github.com/v3/repos/statuses/#get-the-combined-status-for-a-specific-reference
+        if combined_status_state == 'failure':
+            ci_failed = True
+        elif combined_status_state == 'pending':
+            ci_pending = True
+
+        if 'WIP' in pull_request_obj.title:
+            wip_in_title = True
+        work_in_progress_conditions = wip_in_title or pull_request_obj.draft or ci_failed
+        if work_in_progress_conditions:
+            self._add_label(pull_request_obj, 'pr-work-in-progress')
+        elif ci_pending:
+            self._add_label(pull_request_obj, 'pr-awaiting-testing')
+        else:  # CI passed since status=successful
+            # parse reviews to assess count of approved/requested changes/commented/dismissed reviews
+            approves, request_changes, comments, dismissed = self._parse_reviews(pull_request_obj)
+            if approves > 0 and request_changes == 0:
+                self._add_label(pull_request_obj, 'pr-awaiting-merge')
+            else:
+                has_no_reviews = approves + request_changes - dismissed + comments == 0
```

Comment:
I'd like to see some units tests for all these cases. Seems too complex for me, hence please add tests